src/PAModel/Serializers/SourceSerializer.cs

```diff
@@ -795,5 +797,78 @@ private static BuildVerJson GetBuildDetails()
             }
         }
 
+        private static void ValidateIfTemplateExists(CanvasDocument app, BlockNode node, ErrorContainer errors)
+        {
+            // parse all children of the root nodes to see if al of them have the temeplate they require.
+            foreach(var child in node.Children)
+            {
+                if (string.IsNullOrEmpty(child?.Name?.Kind?.TypeName))
+                {
+                    // could not perform validation, hence continue.
+                    continue;
+                }
+
+                // check if it is a group control. group control isn't a real control and don't have templates, but have child control which have templates.
+                if (child.Name.Kind.TypeName == "group" || child.Name.Kind.TypeName == "groupContainer")
+                {
+                    foreach (var child1 in child.Children)
+                    {
+                        ValidateIfTemplateExistsInternal(app, node, child1, errors);
+                    }
+                }
+                // validate if template exists.
+                else
+                {
+                    ValidateIfTemplateExistsInternal(app, node, child, errors);
+                }
+            }
+        }
+
+        private static void ValidateIfTemplateExistsInternal(CanvasDocument app, BlockNode parentNode, BlockNode childNode, ErrorContainer errors)
+        {
+            if (string.IsNullOrEmpty(childNode?.Name?.Kind?.TypeName))
+            {
+                // could not perform validation.
+                return;
+            }
+            // group and grouContainer controls aren't real controls they contain other child controls.
+            if (childNode.Name.Kind.TypeName == "group" || childNode.Name.Kind.TypeName == "groupContainer")
```

Comment:
What about galleries and other nested controls? We still need to validate the child controls for them as well.

Reply:
can I assume that if any control has any children run the validation for all the children, so that I don't have to check for certain specific controls? Currently I was only running the validation on children for group/groupContainer controls, probably I can run for all controls with children.

Reply:
yup, that's a good approach.